File: clients/ios/Lextures/Core/LMS/BoardsAdvancedLogic.swift
import Foundation

/// MOB.8 helpers — templates, export polling, present ordering, governance gating.
enum BoardsAdvancedLogic {
    /// Templates/export/present/analytics require course boards + the mobile advanced gate.
    static func isAdvancedEnabled(
        courseEnabled: Bool,
        features: MobilePlatformFeatures
    ) -> Bool {
        courseEnabled && features.ffMobileBoardsAdvanced
    }

    static func canUseTemplates(
        courseEnabled: Bool,
        features: MobilePlatformFeatures,
        canCreate: Bool
    ) -> Bool {
        isAdvancedEnabled(courseEnabled: courseEnabled, features: features) && canCreate
    }

    static func canExportOrPresent(
        courseEnabled: Bool,
        features: MobilePlatformFeatures,
        canManage: Bool
    ) -> Bool {
        isAdvancedEnabled(courseEnabled: courseEnabled, features: features) && canManage
    }

    static func canViewBoardAnalytics(
        courseEnabled: Bool,
        features: MobilePlatformFeatures,
        canManage: Bool
    ) -> Bool {
        isAdvancedEnabled(courseEnabled: courseEnabled, features: features) && canManage
    }

    static func filterTemplates(
        _ templates: [BoardTemplate],
        scope: BoardTemplateScope?,
        query: String
    ) -> [BoardTemplate] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return templates.filter { template in
            if let scope, template.scope.lowercased() != scope.rawValue { return false }
            guard !needle.isEmpty else { return true }
            let hay = [
                template.title,
                template.description,
                template.tags.joined(separator: " "),
            ].joined(separator: " ").lowercased()
            return hay.contains(needle)
        }
    }

    /// Exponential backoff for export/copy job polling (seconds): 0.5, 1, 2, 4… capped.
    static func pollDelaySeconds(attempt: Int, cap: TimeInterval = 8) -> TimeInterval {
        let base = 0.5 * pow(2.0, Double(max(0, attempt)))
        return min(cap, base)
    }

    static func isExportTerminal(_ status: String) -> Bool {
        let s = status.lowercased()
        return s == "done" || s == "failed"
    }

    static func isCopyTerminal(_ status: String) -> Bool {
        let s = status.lowercased()
        return s == "completed" || s == "failed"
    }

    static func exportFileExtension(format: BoardExportFormat) -> String {
        switch format {
        case .pdf: "pdf"
        case .csv: "csv"
        case .image: "png"
        }
    }

    static func exportMimeType(format: BoardExportFormat) -> String {
        switch format {
        case .pdf: "application/pdf"
        case .csv: "text/csv"
        case .image: "image/png"
        }
    }

    /// Present-mode ordering: section order, then card sortIndex (web present-mode parity).
    static func orderedPostsForPresent(posts: [BoardPost], sections: [BoardSection]) -> [BoardPost] {
        let secOrder = Dictionary(uniqueKeysWithValues: sections.map { ($0.id, $0.sortIndex) })
        return posts.sorted { a, b in
            let aSec = a.sectionId.flatMap { secOrder[$0] } ?? Double.greatestFiniteMagnitude
            let bSec = b.sectionId.flatMap { secOrder[$0] } ?? Double.greatestFiniteMagnitude
            if aSec != bSec { return aSec < bSec }
            return a.sortIndex < b.sortIndex
        }
    }

    static func postBodyText(_ post: BoardPost) -> String {
        if let text = post.body?.text?.trimmingCharacters(in: .whitespacesAndNewlines), !text.isEmpty {
            return text
        }
        if let html = post.body?.html, !html.isEmpty {
            return stripHTML(html)
        }
        let title = post.title.trimmingCharacters(in: .whitespacesAndNewlines)
        return title
    }

    private static func stripHTML(_ html: String) -> String {
        var result = html
        if let regex = try? NSRegularExpression(pattern: "<[^>]+>", options: .caseInsensitive) {
            let range = NSRange(result.startIndex..., in: result)
            result = regex.stringByReplacingMatches(in: result, options: [], range: range, withTemplate: "")
        }
        return result
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func formatStorageBytes(_ bytes: Int64) -> String {
        BoardsLogic.formatFileSize(bytes)
    }

    static func parseBoardCapDraft(_ raw: String) -> Int? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        return Int(trimmed)
    }
}

enum BoardsAdvancedObservability {
    private static var counters: [String: Int] = [:]
    private static let lock = NSLock()

    static func record(_ event: String, attributes: [String: String] = [:]) {
        lock.lock()
        defer { lock.unlock() }
        let key = attributes.isEmpty
            ? event
            : event + "|" + attributes.keys.sorted().map { "\($0)=\(attributes[$0] ?? "")" }.joined(separator: ",")
        counters[key, default: 0] += 1
    }

    static func count(for event: String) -> Int {
        lock.lock()
        defer { lock.unlock() }
        return counters.filter { $0.key == event || $0.key.hasPrefix(event + "|") }.values.reduce(0, +)
    }

    #if DEBUG
    static func resetForTests() {
        lock.lock()
        counters.removeAll()
        lock.unlock()
    }
    #endif
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/BoardsAdvancedLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import java.util.Locale
import kotlin.math.min
import kotlin.math.pow

/** MOB.8 helpers — templates, export polling, present ordering, governance gating. */
object BoardsAdvancedLogic {
    fun isAdvancedEnabled(courseEnabled: Boolean, features: MobilePlatformFeatures): Boolean =
        courseEnabled && features.ffMobileBoardsAdvanced

    fun canUseTemplates(
        courseEnabled: Boolean,
        features: MobilePlatformFeatures,
        canCreate: Boolean,
    ): Boolean = isAdvancedEnabled(courseEnabled, features) && canCreate

    fun canExportOrPresent(
        courseEnabled: Boolean,
        features: MobilePlatformFeatures,
        canManage: Boolean,
    ): Boolean = isAdvancedEnabled(courseEnabled, features) && canManage

    fun canViewBoardAnalytics(
        courseEnabled: Boolean,
        features: MobilePlatformFeatures,
        canManage: Boolean,
    ): Boolean = isAdvancedEnabled(courseEnabled, features) && canManage

    fun filterTemplates(
        templates: List<BoardTemplate>,
        scope: BoardTemplateScope?,
        query: String,
    ): List<BoardTemplate> {
        val needle = query.trim().lowercase(Locale.US)
        return templates.filter { template ->
            if (scope != null && !template.scope.equals(scope.apiValue, ignoreCase = true)) return@filter false
            if (needle.isEmpty()) return@filter true
            val hay = listOf(template.title, template.description, template.tags.joinToString(" "))
                .joinToString(" ")
                .lowercase(Locale.US)
            hay.contains(needle)
        }
    }

    fun pollDelaySeconds(attempt: Int, cap: Double = 8.0): Double {
        val base = 0.5 * 2.0.pow(maxOf(0, attempt).toDouble())
        return min(cap, base)
    }

    fun isExportTerminal(status: String): Boolean {
        val s = status.lowercase(Locale.US)
        return s == "done" || s == "failed"
    }

    fun isCopyTerminal(status: String): Boolean {
        val s = status.lowercase(Locale.US)
        return s == "completed" || s == "failed"
    }

    fun exportFileExtension(format: BoardExportFormat): String = when (format) {
        BoardExportFormat.Pdf -> "pdf"
        BoardExportFormat.Csv -> "csv"
        BoardExportFormat.Image -> "png"
    }

    fun exportMimeType(format: BoardExportFormat): String = when (format) {
        BoardExportFormat.Pdf -> "application/pdf"
        BoardExportFormat.Csv -> "text/csv"
        BoardExportFormat.Image -> "image/png"
    }

    fun orderedPostsForPresent(posts: List<BoardPost>, sections: List<BoardSection>): List<BoardPost> {
        val secOrder = sections.associate { it.id to it.sortIndex }
        return posts.sortedWith(
            compareBy<BoardPost> { post ->
                post.sectionId?.let { secOrder[it] } ?: Double.POSITIVE_INFINITY
            }.thenBy { it.sortIndex },
        )
    }

    fun postBodyText(post: BoardPost): String {
        val text = post.body?.text?.trim().orEmpty()
        if (text.isNotEmpty()) return text
        val html = post.body?.html.orEmpty()
        if (html.isNotEmpty()) {
            return html
                .replace(Regex("<[^>]+>"), "")
                .replace("&nbsp;", " ")
                .replace("&amp;", "&")
                .replace("&lt;", "<")
                .replace("&gt;", ">")
                .trim()
        }
        return post.title.trim()
    }

    fun formatStorageBytes(bytes: Long): String = BoardsLogic.formatFileSize(bytes)

    fun parseBoardCapDraft(raw: String): Int? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        return trimmed.toIntOrNull()
    }
}

object BoardsAdvancedObservability {
    private val counters = mutableMapOf<String, Int>()
    private val lock = Any()

    fun record(event: String, attributes: Map<String, String> = emptyMap()) {
        synchronized(lock) {
            val key = if (attributes.isEmpty()) {
                event
            } else {
                event + "|" + attributes.keys.sorted().joinToString(",") { "$it=${attributes[it]}" }
            }
            counters[key] = (counters[key] ?: 0) + 1
        }
    }

    fun count(event: String): Int = synchronized(lock) {
        counters.filter { it.key == event || it.key.startsWith("$event|") }.values.sum()
    }

    fun resetForTests() {
        synchronized(lock) { counters.clear() }
    }
}
